### PR TITLE
feat(logging): Add proper VSCode Logging Output

### LINKED ISSUE #227: bazel query getting stuck
The Bazel extension occasionally gets stuck. I think it usually happens when bazel crashes once, and leaves the lock file around. I start getting errors that continuously pop up, which look like this:

```
Bazel Build Targets: Command failed: bazel --output_base=/tmp/6a622852f8f9062446271ea7e15aa515 query ...:* --output=package 
Another command holds the client lock: pid=58100 owner=client cwd=/workspace/___
Waiting for it to complete... 
Another command holds the client lock: pid=58342 owner=client cwd=/workspace/___
Waiting for it to complete... 
...
 WARNING: Waiting for server process to terminate (waited 5 seconds, waiting at most 10)
WARNING: Waiting for server process to terminate (waited 10 seconds, waiting at most 10)
INFO: Waited 10 seconds for server process (pid=42461) to terminate.
FATAL: Attempted to kill stale server process (pid=42461) using SIGKILL, but it did not die in a timely fashion.
```

I tried deleting /tmp/616.../lock, but the errors continue showing up.

And I have a lot of bazel processes that keep changing PIDs:

```bash
$ ps aux | grep bazel | wc -l
524
```

For now, I really just need syntax highlighting. Is there a way to disable the continuous queries until this issues is fixed?

### PR DESCRIPTION
**Pull request by markkohdev**
## Description

This PR introduces a centralized logging module to replace direct `console.log`/`console.error` calls and scattered `vscode.window.show*Message` usage throughout the extension. The new logger provides:

- **Structured logging** via VS Code's `LogOutputChannel` with proper log levels (error, warn, info, debug)
- **Optional user-facing messages** with a "Details" button that opens the output channel for full context
- **Consistent error formatting** with support for additional arguments using `util.format`
- **Better debugging experience** by centralizing all logs in a single output channel

### Key Changes

- Added `src/extension/logger.ts` with logging API (`logError`, `logWarn`, `logInfo`, `logDebug`)
- Refactored 15+ files to use the new logger instead of direct console/window message calls
- Improved error handling in `bazel_query.ts` (better error formatting, fixed duplicate query error)
- Enhanced `bazel_utils.ts` error handling for invalid regex patterns and missing files
- Added comprehensive test suite (`test/logger.test.ts`) with 573 lines covering all logging scenarios

### Usage Examples

**Setup (in extension activation):**
```typescript
import { registerLogger, logInfo, logError } from "./extension/logger";

export async function activate(context: vscode.ExtensionContext) {
  registerLogger("Bazel VSCode", context);
  logInfo("Extension activated successfully.");
  // ...
}
```

**Basic logging (output channel only):**
```typescript
import { logDebug, logInfo, logWarn, logError } from "./extension/logger";

logDebug("Running Bazel query", false, `Command: ${command}`);
logInfo("Build completed successfully");
logWarn("Buildifier not found", false);
logError("Query failed", false, error);
### Usage Examples

**Setup (in extension activation):**
```typescript
import { registerLogger, logInfo, logError } from "./extension/logger";

export async function activate(context: vscode.ExtensionContext) {
  registerLogger("Bazel VSCode", context);
  logInfo("Extension activated successfully.");
  // ...
}
```

**Basic logging (output channel only):**
```typescript
import { logDebug, logInfo, logWarn, logError } from "./extension/logger";

logDebug("Running Bazel query", false, `Command: ${command}`);
logInfo("Build completed successfully");
logWarn("Buildifier not found", false);
logError("Query failed", false, "Error details:", error);
logError("Package discovery failed", false, "Unable to find file: %s", filepath);
```

**User-facing messages with details:**
```typescript
// Shows error notification with "Details" button
logError("Failed to find Bazel workspace", true);

// Shows info notification + logs full details to output channel
logInfo("Bazel build completed successfully in 5 seconds.", true, buildOutputStr);

// Shows warning + logs additional context
logWarn(
  "Buildifier not available; linting disabled.",
  true,
  "Download from:",
  BUILDTOOLS_RELEASES_URL
);
```

**Error handling with context:**
```typescript
try {
  await bazelQuery.execute();
} catch (error) {
  logError(
    "Bazel query failed",
    true,  // Show user notification
    "Query command output:",
    errorOutput,
    "Exit code:",
    code
  );
}
```

### Additional Improvements

- Added PR template (`.github/pull_request_template.md`) for consistent PR descriptions
- Updated `CONTRIBUTING.md` with clearer testing instructions
- Added `npm run clean` script for easier test environment cleanup
- Removed `out/` dir from vscode file exclusion to facilitate easier test debugging

## Related Issue

Closes #227 

## Testing

- [x] Tests added/updated
  - Added comprehensive unit tests for logger module (573 lines)
  - Tests cover all log levels, user messaging, argument formatting, and edge cases
  - All existing tests continue to pass
- [x] Manual testing performed
  - Verified logging output appears correctly in VS Code output channel
  - Confirmed user messages display with "Details" button functionality
  - Tested error scenarios to ensure proper logging and user notifications

## Checklist

- [x] Code follows the project's style guidelines (prettier/eslint)
- [x] Commit messages follow [Conventional Commit](https://www.conventionalcommits.org/) conventions
- [x] Tests pass locally (`npm run test`)

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -1,7 +1,6 @@
 {
     "files.exclude": {
-        "node_modules": true,
-        "out": true
```
Comment (markkohdev) [inline .vscode/settings.json:4]: Note: Excluding the `out` directory from the `files` panel made testing very difficult since there wasn't visibility into stale files (which I needed to clean up with the new `npm run clean` command).  If there were particular reasons that this was excluded, then we should use a more targeted approach than just `files.exclude`

File: src/bazel/bazel_utils.ts [hunk 2]
```diff
@@ -111,8 +111,13 @@ function findAncestorFile(
   let iteration = 0;
   const maxIterations = 100; // Fail-safe to prevent infinite loops
 
-  if (fs.statSync(startPath).isFile()) {
-    dirname = path.dirname(dirname);
+  try {
```
Comment (markkohdev) [inline src/bazel/bazel_utils.ts:118]: Macs don't have some of the files that are searched for during quickpick tests ("/etc/hostname"), causing crashing when trying to run tests locally.

Comment (cbandera) [reply src/bazel/bazel_utils.ts:118]: I wonder whether we shouldn't rather adapt the testcase then

File: src/extension/logger.ts [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+// https://github.com/microsoft/vscode-black-formatter/blob/main/src/common/logging.ts
+import * as util from "util";
+import * as vscode from "vscode";
+import { Disposable, ExtensionContext, LogOutputChannel } from "vscode";
+
+type Arguments = unknown[];
+
+const DETAILS_ACTION = "Details";
+let channel: LogOutputChannel | undefined;
+let outputChannelName: string | undefined;
```
Comment (cbandera) [inline src/extension/logger.ts]: I would prefer to not make the channel name configurable but hardcode it here.
Also I would call it `Bazel` instead of `Bazel Vscode` (the vscode in the repo name is really just convention for naming extensions... but user facing, `Bazel` is what makes most sense

File: src/extension/logger.ts [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+      log: true,
+    },
+  );
+  context.subscriptions.push(outputChannel);
+
+  outputChannelName = name;
+  channel = outputChannel;
+  channel.info("Logger initialized.");
+
+  const loggerDisposable: Disposable = {
```
Comment (cbandera) [inline src/extension/logger.ts:56]: `createOutputChannel` already returns a `dispose`able object. No need to wrap it again

File: src/extension/logger.ts [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+}
+
+/**
+ * Shows the output channel in the VS Code Output panel.
+ * This is useful for displaying error messages with a link to open the output.
+ */
+export function showOutputChannel(): void {
+  if (channel) {
+    channel.show(true);
+  } else if (outputChannelName) {
```
Comment (cbandera) [inline src/extension/logger.ts]: We don't need the `else` branch here. Creating a channel should be the pure responsibility of `registerLogger`

File: src/test-explorer/lcov_parser.ts [hunk 6]
```diff
@@ -123,7 +123,9 @@ async function demangleNameUsingFilter(
   mangled: string,
 ): Promise<string | undefined> {
   if (execPath === null) return undefined;
-  const unmangled = (await execFile(execPath, [mangled])).stdout.trim();
+  // c++filt needs the -n flag to properly demangle names that start with underscore
```
Comment (cbandera) [inline src/test-explorer/lcov_parser.ts:126]: This change seems unrelated. Might this also be something mac-related? The tests have worked as-is for me.

File: test/logger.test.ts [hunk 7]
```diff
@@ -0,0 +1,573 @@
...
+    );
+    assert.ok(debugCall, "logVerbose should call debug");
+    assert.strictEqual(
+      debugCall.message,
+      "verbose message",
+      "logVerbose should format message correctly",
+    );
+  });
+
+  it("handles dispose correctly", () => {
```
Comment (cbandera) [inline test/logger.test.ts:219]: Let's also assert that the global vars are set back to undefined

File: src/extension/logger.ts [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+  ...args: Arguments
+): void {
+  if (!channel) {
+    return;
+  }
+
+  // Log the detailed message to the output channel
+  let detailedMessage = message;
+  if (args.length > 0) {
+    detailedMessage += " Details:\n" + util.format(...args);
```
Comment (cbandera) [inline src/extension/logger.ts:89]:
```suggestion
    detailedMessage += ":\n" + util.format(...args);
```
I feel the additional Details word is just confusing here:
> [debug] Running Bazel query Details:

File: src/extension/bazel_wrapper_commands.ts [hunk 9]
```diff
@@ -431,7 +429,7 @@ async function bazelGoToLabel(target_info?: blaze_query.ITarget | undefined) {
 function copyLabelToClipboard(label: string): void {
   vscode.env.clipboard.writeText(label);
 
-  vscode.window.showInformationMessage(`Copied to clipboard: ${label}`);
+  logInfo(`Copied to clipboard: ${label}`, true);
```
Comment (cbandera) [inline src/extension/bazel_wrapper_commands.ts]: This message does also show the `Details` button, but there is really not any other info to see in the logging channel. 
I wonder whether we show introduce another flag to configure whether to show the button or not.

**Review by cbandera — COMMENTED** [review]
Thank you for this great contribution! It's a feature that has already been requested several times.
I have done a first review and left you some findings.
In addition, I would like to ask you to consider introducing a new VSCode Command  `Bazel: Show output`. This way a user can always quickly bring up the output channel to view what's currently going on. An example implementation can be found [here](https://github.com/bazel-contrib/vscode-bazel/pull/496/changes/84fa7c4d1c1685d0c12d8011b73f8505b091bf6f#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519)

File: src/extension/bazel_wrapper_commands.ts (hunk 9 above)
Comment (markkohdev) [reply src/extension/bazel_wrapper_commands.ts]: That makes sense 👍 
I think that if there's not really any details to show then we can just show the message without emitting a log.  Rather than trying to propagate something like `showDetailsButton: bool` all the way through the log methods I'll just add the extra param with a default to the `showUserMessage` function.

File: src/extension/logger.ts (hunk 3 above)
Comment (markkohdev) [reply src/extension/logger.ts]: Fair enough, I was trying to make this somewhat generic so that we could potentially add additional output channels (while still maintaining a global default) in the future if the extension scope grew (similar to having multiple logger classes in large applications), but in honesty it was a sort of half-baked effort that I didn't follow-through with since I made all of the logger methods reference the global default haha 😅  
I think it makes sense to be opinionated here for now.  If the need truly arises in the future to create a multi-channel logging system then we can address it then.

File: src/extension/logger.ts (hunk 8 above)
Comment (markkohdev) [reply src/extension/logger.ts:89]: I found that it helped for some messages, but I don't have a strong attachment to it so I can remove

File: src/extension/logger.ts (hunk 5 above)
Comment (markkohdev) [reply src/extension/logger.ts]: Yep, makes sense with the hard-coded channel.

File: src/test-explorer/lcov_parser.ts (hunk 6 above)
Comment (markkohdev) [reply src/test-explorer/lcov_parser.ts:126]: yeah, I think this may be specifically due to the c++filt and rustfilt implementations for MacOS.  I take it you're not on a mac?

File: src/extension/logger.ts (hunk 4 above)
Comment (markkohdev) [reply src/extension/logger.ts:56]: The disposable wrapper was so that we could clear the globals.  Any suggestion for how to do that without the wrapping?  (typescript is definitely not my strong suit)

**Comment by markkohdev** [comment]
@cbandera thanks for the review!  I've addressed most of your comments but let me know if you want a different approach for the disposable.

File: src/extension/logger.ts (hunk 8 above)
Comment (cbandera) [reply src/extension/logger.ts:89]: NVM. I don't have a strong opinion on it either.

File: src/extension/logger.ts (hunk 4 above)
Comment (cbandera) [reply src/extension/logger.ts:56]: I see, I didn't see the connection. But that's fine then.

File: src/test-explorer/lcov_parser.ts (hunk 6 above)
Comment (cbandera) [reply src/test-explorer/lcov_parser.ts:126]: Yes. I am on linux.
